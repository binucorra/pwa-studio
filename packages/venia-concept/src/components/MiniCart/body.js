import React, { Fragment, useCallback, useState } from 'react';
import { bool, func, object, shape, string } from 'prop-types';

import { useItemBeingEdited } from '@magento/peregrine';

import EmptyMiniCart from './emptyMiniCart';
import Footer from './footer';
import ProductList from './productList';

const Body = props => {
    // Props.
    const {
        cart,
        classes,
        isCartEmpty,
        isMiniCartMaskOpen,
        openOptionsDrawer,
        removeItemFromCart
    } = props;

    if (isCartEmpty) {
        return <EmptyMiniCart />;
    }

    // State.
    const [_, setItemBeingEdited] = useItemBeingEdited();

    // Members.
    const cartId = cart.details.id;
    const currencyCode = cart.details.currency.quote_currency_code;

    // Callbacks.
    const handleOpenOptionsDrawer = useCallback(
        item => {
            console.log('setting item being edited to', item);
            setItemBeingEdited(item);
            openOptionsDrawer();
        },
        [openOptionsDrawer, setItemBeingEdited]
    );

    return (
        <Fragment>
            <div className={classes.body}>
                {cartId && (
                    <ProductList
                        currencyCode={currencyCode}
                        items={cart.details.items}
                        openOptionsDrawer={handleOpenOptionsDrawer}
                        removeItemFromCart={removeItemFromCart}
                        totalsItems={cart.totals.items}
                    />
                )}
            </div>
            <Footer
                cart={cart}
                classes={classes}
                isMiniCartMaskOpen={isMiniCartMaskOpen}
            />
        </Fragment>
    );
};

Body.propTypes = {
    cart: shape({
        details: object,
        totals: object
    }),
    classes: shape({
        footer: string,
        footerMaskOpen: string,
        summary: string
    }),
    isCartEmpty: bool,
    isMiniCartMaskOpen: bool,
    openOptionsDrawer: func.isRequired,
    removeItemFromCart: func
};

export default Body;
