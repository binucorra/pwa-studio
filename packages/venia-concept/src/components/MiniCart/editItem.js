import React, { useEffect } from 'react';
import { bool, func, object } from 'prop-types';

import { useQuery } from '@magento/peregrine';

import { loadingIndicator } from 'src/components/LoadingIndicator';
import PRODUCT_DETAILS from 'src/queries/getProductDetailByName.graphql';

import CartOptions from './cartOptions';

const EditItem = props => {
    // Props.
    const { endEditItem, isUpdatingItem, item, updateItemInCart } = props;

    // We must have an item to edit.
    if (!item) {
        // TODO: log an error?
        return null;
    }

    // State / Hooks.
    const [queryResult, queryApi] = useQuery(PRODUCT_DETAILS);
    const { data, error, loading } = queryResult;
    const { runQuery, setLoading } = queryApi;

    // Members.
    const itemHasOptions = item.options.length > 0;

    // Don't have to run the query for non-configurable items.
    if (!itemHasOptions) {
        return (
            <CartOptions
                cartItem={item}
                configItem={{}}
                endEditItem={endEditItem}
                isUpdatingItem={isUpdatingItem}
                updateCart={updateItemInCart}
            />
        );
    }

    // Run the query once on mount and again whenever the
    // item being edited changes.
    useEffect(() => {
        setLoading(true);
        runQuery({
            variables: {
                name: item.name,
                onServer: false
            }
        });
    }, [item]);

    /*
     * Render.
     */

    // We don't have data yet, we're either loading it or
    // in an error situation.
    if (!data) {
        let status;
        if (error) status = <span>Unable to fetch item options</span>;
        if (loading) status = loadingIndicator;

        return <span>{status}</span>;
    }

    // We do have this item's data.
    const itemWithOptions = data.products.items[0];

    return (
        <CartOptions
            cartItem={item}
            configItem={itemWithOptions}
            endEditItem={endEditItem}
            isUpdatingItem={isUpdatingItem}
            updateCart={updateItemInCart}
        />
    );
};

EditItem.propTypes = {
    endEditItem: func,
    isUpdatingItem: bool,
    item: object,
    updateItemInCart: func
};

export default EditItem;
