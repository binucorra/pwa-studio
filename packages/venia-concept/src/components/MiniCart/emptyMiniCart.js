import React from 'react';
import { string, shape } from 'prop-types';

import { mergeClasses } from 'src/classify';
import Trigger from './trigger';
import defaultClasses from './emptyMiniCart.css';

const EmptyMiniCart = props => {
    const classes = mergeClasses(defaultClasses, props.classes);

    return (
        <div className={classes.root}>
            <h3 className={classes.emptyTitle}>
                There are no items in your shopping cart
            </h3>
            <Trigger>
                <span className={classes.continue}>Continue Shopping</span>
            </Trigger>
        </div>
    );
};

EmptyMiniCart.propTypes = {
    classes: shape({
        root: string,
        emptyTitle: string,
        continue: string
    })
};

export default EmptyMiniCart;
