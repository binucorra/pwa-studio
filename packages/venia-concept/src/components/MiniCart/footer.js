import React, { Suspense } from 'react';
import { bool, object, shape, string } from 'prop-types';

import { mergeClasses } from 'src/classify';
import Checkout from 'src/components/Checkout';
import CheckoutButton from 'src/components/Checkout/checkoutButton';

import defaultClasses from './footer.css';
import TotalsSummary from './totalsSummary';

const Footer = props => {
    const { cart, isCartEmpty, isMiniCartMaskOpen } = props;
    const { isEditingItem, isLoading } = cart;

    const classes = mergeClasses(defaultClasses, props.classes);
    const footerClassName = isMiniCartMaskOpen
        ? classes.root_open
        : classes.root;
    const placeholderButton = (
        <div className={classes.placeholderButton}>
            <CheckoutButton ready={false} />
        </div>
    );

    if (isCartEmpty || isLoading || isEditingItem) {
        return null;
    }

    return (
        <div className={footerClassName}>
            <TotalsSummary cart={cart} />
            <Suspense fallback={placeholderButton}>
                <Checkout cart={cart} />
            </Suspense>
        </div>
    );
};

Footer.propTypes = {
    cart: shape({
        isEditingItem: bool,
        isLoading: bool
    }).isRequired,
    classes: shape({
        placeholderButton: string,
        root: string,
        root_open: string,
        summary: string
    }),
    isCartEmpty: bool,
    isMiniCartMaskOpen: bool
};

export default Footer;
