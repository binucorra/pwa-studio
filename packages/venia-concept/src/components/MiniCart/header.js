import React from 'react';
import { bool, shape, string } from 'prop-types';
import CloseIcon from 'react-feather/dist/icons/x';

import { mergeClasses } from 'src/classify';
import Icon from 'src/components/Icon';

import defaultClasses from './header.css';
import Trigger from './trigger';

const Header = props => {
    const { cart } = props;

    const classes = mergeClasses(defaultClasses, props.classes);
    const { isEditingItem } = cart;
    const title = isEditingItem ? 'Edit Cart Item' : 'Shopping Cart';

    return (
        <div className={classes.root}>
            <h2 className={classes.title}>{title}</h2>
            <Trigger>
                <Icon src={CloseIcon} />
            </Trigger>
        </div>
    );
};

Header.propTypes = {
    cart: shape({
        isEditingItem: bool
    }),
    classes: shape({
        root: string,
        title: string
    })
};

export default Header;
