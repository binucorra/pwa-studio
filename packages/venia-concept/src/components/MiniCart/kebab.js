import React, { useCallback, useRef, useState } from 'react';
import { node, shape, string } from 'prop-types';
import MoreVerticalIcon from 'react-feather/dist/icons/more-vertical';
import { useEventListener } from '@magento/peregrine';

import { mergeClasses } from 'src/classify';
import Icon from 'src/components/Icon';

import defaultClasses from './kebab.css';

const Kebab = props => {
    const { children } = props;

    const kebabButtonRef = useRef(null);
    const [isOpen, setIsOpen] = useState(false);

    const classes = mergeClasses(defaultClasses, props.classes);

    const handleDocumentClick = useCallback(
        event => {
            const shouldBeOpen = kebabButtonRef.current.contains(event.target);
            setIsOpen(shouldBeOpen);
        },
        [kebabButtonRef, setIsOpen]
    );

    useEventListener(document, 'mousedown', handleDocumentClick);
    useEventListener(document, 'touchend', handleDocumentClick);

    const toggleClass = isOpen ? classes.dropdown_active : classes.dropdown;

    return (
        <div className={classes.root}>
            <button className={classes.kebab} ref={kebabButtonRef}>
                <Icon src={MoreVerticalIcon} />
            </button>
            <ul className={toggleClass}>{children}</ul>
        </div>
    );
};

Kebab.propTypes = {
    children: node,
    classes: shape({
        dropdown: string,
        dropdown_active: string,
        kebab: string,
        root: string
    })
};

export default Kebab;
