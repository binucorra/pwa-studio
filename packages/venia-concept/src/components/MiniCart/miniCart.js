import React from 'react';
import { bool, func, object, shape, string } from 'prop-types';

import Body from './body';
import Header from './header';

import { mergeClasses } from 'src/classify';
import defaultClasses from './miniCart.css';

const MiniCart = props => {
    // Props.
    const {
        beginEditItem,
        cart,
        endEditItem,
        isCartEmpty,
        isMiniCartMaskOpen,
        isOpen,
        removeItemFromCart,
        updateItemInCart
    } = props;

    // Members.
    const classes = mergeClasses(defaultClasses, props.classes);
    const rootClass = isOpen ? classes.root_open : classes.root;

    return (
        <aside className={rootClass}>
            <Header cart={cart} />
            <Body
                beginEditItem={beginEditItem}
                cart={cart}
                endEditItem={endEditItem}
                isCartEmpty={isCartEmpty}
                isMiniCartMaskOpen={isMiniCartMaskOpen}
                removeItemFromCart={removeItemFromCart}
                updateItemInCart={updateItemInCart}
            />
        </aside>
    );
};

MiniCart.propTypes = {
    beginEditItem: func.isRequired,
    endEditItem: func.isRequired,
    cart: object,
    classes: shape({
        header: string,
        root: string,
        root_open: string,
        title: string
    }),
    isMiniCartMaskOpen: bool,
    removeItemFromCart: func,
    updateItemInCart: func
};

export default MiniCart;
