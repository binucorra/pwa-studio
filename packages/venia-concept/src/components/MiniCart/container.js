import { connect } from 'src/drivers';
import { isEmptyCartVisible, isMiniCartMaskOpen } from 'src/selectors/cart';
import {
    beginEditItem,
    endEditItem,
    getCartDetails,
    updateItemInCart,
    removeItemFromCart
} from 'src/actions/cart';
import { cancelCheckout } from 'src/actions/checkout';

import MiniCart from './miniCart';

const mapStateToProps = state => {
    const { cart } = state;

    return {
        cart,
        isCartEmpty: isEmptyCartVisible(state),
        isMiniCartMaskOpen: isMiniCartMaskOpen(state)
    };
};

const mapDispatchToProps = {
    beginEditItem,
    cancelCheckout,
    endEditItem,
    getCartDetails,
    removeItemFromCart,
    updateItemInCart
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(MiniCart);
