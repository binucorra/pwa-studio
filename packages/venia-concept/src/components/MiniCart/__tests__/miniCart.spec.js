import React from 'react';
import { createTestInstance } from '@magento/peregrine';

import MiniCart from '../miniCart';

jest.mock('../body');
jest.mock('../header');
jest.mock('../footer');

test('renders correctly', () => {
    const props = {
        beginEditItem: jest.fn(),
        endEditItem: jest.fn(),
        cart: {
            isEditingItem: false
        }
    };

    const tree = createTestInstance(<MiniCart {...props} />).toJSON();

    expect(tree).toMatchSnapshot();
});
