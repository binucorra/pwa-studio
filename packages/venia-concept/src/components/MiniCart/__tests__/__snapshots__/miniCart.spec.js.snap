// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<aside>
  ( Header Component Here )
  ( Body Component Here )
  ( Footer Component Here )
</aside>
`;
